<!DOCTYPE html
	PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
	 "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US" xml:lang="en-US">
<head>
<title>HVDR Mind the Gap</title>
<link rel="stylesheet" type="text/css" href="/res/hvdr.css"/>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1"/>
<style type="text/css">
.hidden { visibility: hidden; }
.unhidden { visibility: visible; }
td { text-align : left; }
td.shade1 { background: lightblue; }
td.shade2 { background: #ffff99;w; }
td.shade3 { background: pink; }
fieldset { width: 70%; }
input.aminoacid1 { text-align:center; background:lightblue; color:darkslategrey; font-weight:bold; }
input.aminoacid2 { text-align:center; background:#ffff99; color:darkslategrey; font-weight:bold; }
input.nucleotide { text-align:center; }
div.instructions { font-family:monospace; }
a:focus { outline: 0; }
em.i { font-style:oblique; font-weight:normal; text-decoration:none; }
em.b { font-style:normal; font-weight:bold; text-decoration:none; }
legend { font-weight:bold; padding:0.2em 0.5em; border:1px solid green; color:green; font-size:120%; font-family:monospace; background:lightgreen; }
</style>

<script language="JavaScript">
function RegExpHelp()
{
    alert("Ensure sequence data is *really* in FASTA format.");
}

function showHide(divID)
{
	for (i=0; i<5; i++)
	{
		var item = document.getElementById(divID + i);
		if (item)
		{
			item.className=(item.className=='hidden')?'unhidden':'hidden';
		}
	}
}
</script>
</head>
<body>
<h1>HVDR Online Tools</h1>
<div class="note">Hepatitis Virus Diversity Research <a href="http://www.wits.ac.za/health/hvdrp">Programme</a>, University of the Witwatersrand</div>
<h2><img src="/res/MindTheGap.png" style="vertical-align:middle"> Mind the Gap</h2>

<form method="post" action="/cgi-bin/MindTheGap.cgi" enctype="multipart/form-data">

<div class="intro">This tool splits an input FASTA file into two files, based on the percentage of gaps present at one or more columns in the alignment. If the number of gaps in a column exceeds the threshold (percentage) specified below, all sequences with the gap will be placed into one file (the gaps will be removed first) , and all sequences containing the residue ("insertion") will be placed into a second file. This is repeated for all remaining sequences, with sequences being added to the first to files, so that only two output files are produced.</div>

<br>
<fieldset>
<legend>Specify Input Parameters</legend>

	<table cellpadding="3">
	<tr><td>Aligned FASTA file</td><td><input type="file" name="infile" size="32"></td></tr>
	<tr><td>Gap Percentage Threshold</td><td><input type="text" value="80" name="threshold" size="3">%</td></tr>
</table>
</fieldset>

<br>
<input type="submit" name="submit;" value="Execute!"/></form>

<!--#include virtual="/hvdrp/footer.html" -->

<br>
<div class="dotted_red">
<p class="compact">
&bull; ALL TOOLS ARE PROVIDED WITHOUT ANY WARRANTY &bull; CHECK ALL RESULTS CAREFULLY &bull;
</p>
<p class="compact">
Queries or comments to ...
<!-- Professor Anna Kramvis [Anna.Kramvis (@) wits.ac.za] -->
</p>
</div>
<br>

<div class="dotted_blue">
TUTORIAL<br>Given the following input file and a threshold of 50%:<br>
<br>
>one<br>
AAA<br>
>two<br>
A-A<br>
>three<br>
A-A<br>
</p>
<br>
The tool will produce an output file containing sequences "two" and "three" (no insertions) and a second file containing sequence "one" (insertion).<br>

<br>Given the following input file and a threshold of 40%:<br>
<br>
>one<br>
AAA<br>
>two<br>
A-A<br>
>three<br>
A-A<br>
>four<br>
CC-<br>
>five<br>
CC-<br>

</p>
<br>
The tool will produce an output file containing all five sequences (no insertions) and a second file containing no sequences (insertion).<br>


</body>
</html>

